### PR TITLE
[Feature] Optimize dist_graph/_split_even_to_part memory usage

### PR DESCRIPTION
**Pull request by JingchengYu94**
## Description
According to [this issue](https://github.com/dmlc/dgl/issues/3107), this PR will make the memory consumption of edge_split and node_split decrease when number of worker increases. In my case (graph with 13 billion edges), the memory consumption of edge_split decreases from ~108GB to ~15GB.

@zheng-da Could you please take a look at this PR?

## Checklist
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [$CATEGORY] (such as [NN], [Model], [Doc], [Feature]])
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage
- [x] Code is well-documented
- [x] To the my best knowledge, examples are either not affected by this change,
      or have been fixed to be compatible with this change
- [x] Related issue is referred in this PR
- [ ] If the PR is for a new model/paper, I've updated the example index [here](../examples/README.md).

## Changes
* Add API count_nonzero in dgl backend, to count the number of nonzero element in a tensor
* Add API add in dgl backend, to addup a tensor with a scalar (element-wise)
* Change the process of split_even
  * Old: compute nonzero tensor over whole elements, then compute the offset and take the slice
  * New: count the number of nonzero elements, compute offset, iterate through whole elements by blocks, compute nonzero tensor of each block, and concatenate to final answer. Block size is set to be #elements / #partitions, so it will decrease with worker increases.
* Note: If elements is a dist_tensor, whole elements will still be pulled as before. We need to implement count_nonzero for kvstore before applying this change to the case of dist_tensor. We can discuss about this further.

### CODE REVIEW

**Comment by BarclayII** [comment]
We support PyTorch as old as 1.5.0 so I would recommend you use numpy's `count_nonzero` for PyTorch as well.

File: python/dgl/backend/backend.py [hunk 1]
```diff
@@ -1125,6 +1125,38 @@ def replace_inf_with_zero(x):
...
+        The tensor to be counted
+
+    Returns
+    -------
+    Integer
+        The result 
+    """
+    pass
+
+def add(input, other):
```
Comment (BarclayII) [inline python/dgl/backend/backend.py:1143]: I think you can directly write `+` instead of defining this new function.

Comment (JingchengYu94) [reply python/dgl/backend/backend.py]: Good suggestion. Updated.

**Comment by JingchengYu94** [comment]
@zheng-da Could you please take a look at this PR?

**Review by BarclayII — APPROVED** [review]

File: python/dgl/backend/pytorch/tensor.py [hunk 2]
```diff
@@ -290,6 +291,10 @@ def clamp(data, min_val, max_val):
 def replace_inf_with_zero(x):
     return th.masked_fill(x, th.isinf(x), 0)
 
+def count_nonzero(input):
+    # TODO: fallback to numpy for backward compatibility
+    return np.count_nonzero(input)
```
Comment (zheng-da) [inline python/dgl/backend/pytorch/tensor.py:296]: do we not need to convert it into numpy array first?

File: python/dgl/distributed/dist_graph.py [hunk 3]
```diff
@@ -1075,26 +1075,53 @@ def _even_offset(n, k):
...
-
-    # compute the offset of each split and ensure that the difference of each partition size
-    # is 1.
-    offsets = _even_offset(len(eles), partition_book.num_partitions())
-    assert offsets[-1] == len(eles)
-
-    # Get the elements that belong to the partition.
-    partid = partition_book.partid
-    part_eles = eles[offsets[partid] : offsets[partid + 1]]
+        elements = F.tensor(elements)
```
Comment (zheng-da) [inline python/dgl/distributed/dist_graph.py:1095]: here `elements` is a vector stored in a local machine?
if we can store the entire `elements` array in the local machine, why do we still have memory issue?

File: python/dgl/backend/pytorch/tensor.py (hunk 2 above)
Comment (JingchengYu94) [reply python/dgl/backend/pytorch/tensor.py:296]: It will convert implicitly.

Comment (BarclayII) [reply python/dgl/backend/pytorch/tensor.py:296]: You don't have to.  numpy functions can work directly on PyTorch tensors.

File: python/dgl/distributed/dist_graph.py (hunk 3 above)
Comment (JingchengYu94) [reply python/dgl/distributed/dist_graph.py:1095]: elements is a boolean mask to split train, validate and test sets (I learnt how to use it from this [example](https://github.com/dmlc/dgl/blob/master/examples/pytorch/graphsage/experimental/train_dist_unsupervised.py#L429-L430)). I agree this is not the best design. 
But in this case, elements is boolean tensor, so it cost 1/8 memory of the nonzero_1d tensor, so at least my machines have enough memory for it.
Do you have any plan about redesign the data split part to further reduce the memory cost?

Comment (BarclayII) [reply python/dgl/distributed/dist_graph.py:1095]: `elements` is a boolean vector so it consumes 8x less memory than an int64 tensor.  `nonzero` will give you an int64 tensor whose length is roughly the same order of magnitude.

See also https://discuss.dgl.ai/t/edge-split-consumes-too-much-memory/2130.

Comment (zheng-da) [reply python/dgl/distributed/dist_graph.py:1095]: i see. then my general comment is that we can unify the code for DistTensor and local tensor.

Comment (JingchengYu94) [reply python/dgl/distributed/dist_graph.py:1095]: Yes. But we need to implement a distributed count_nonzero method first. I can do it next week, what do you think?
